assembly/pallets/timestamp/assembly/timestamp.ts

```diff
@@ -0,0 +1,157 @@
+     */
+    public static readonly PREFIX: u8 = 11;
+
+    /**
+     * index of the timestamp module
+     */
+    public static readonly MODULE_INDEX: u8 = 1;
+
+    /**
+     * @description Toggles the current value of didUpdate
+     */
+    static toggleUpdate(): void {
+        const didUpdate = Storage.get(Timestamp.SCALE_TIMESTAMP_DID_UPDATE);
+        const didUpdateValue: Bool = didUpdate.isSome() ? BytesReader.decodeInto<Bool>((<ByteArray>didUpdate.unwrap()).values) : new Bool(false);
+        if(didUpdateValue.value){
+            const falseu8 = new Bool(false);
+            Storage.set(Timestamp.SCALE_TIMESTAMP_DID_UPDATE, falseu8.toU8a());
+        }
+        else{
+            const trueu8 = new Bool(true);
+            Storage.set(Timestamp.SCALE_TIMESTAMP_DID_UPDATE, trueu8.toU8a());
+        }
+    }
+
+    /**
+     * @description Sets the current time. When setting the new time, 
+     * it must be greater than the last one (set into storage) with at least a MinimumPeriod
+     * @param now timestamp number
+     */
+    static set(now: Moment): u8[] {
+        const didUpdate = Storage.get(Timestamp.SCALE_TIMESTAMP_DID_UPDATE);
+        const didUpdateValue: Bool = didUpdate.isSome() ? BytesReader.decodeInto<Bool>((<ByteArray>didUpdate.unwrap()).values) : new Bool(false);
+        if(didUpdateValue.value){
+            Log.error('Validation error: Timestamp must be updated only once in the block');
+            return this._tooFrequentResponseCode();
+        }
+        let minValue = Timestamp.get().value + TimestampTypes.minimumPeriod().value;
+        if(now.value < minValue){
+            Log.error('Validation error: Timestamp must increment by at least <MinimumPeriod> between sequential blocks');
+            return this._timeframeTooLowResponceCode();
+        }
+
+        const trueu8 = new Bool(true);
+        Storage.set(Timestamp.SCALE_TIMESTAMP_DID_UPDATE, trueu8.toU8a());
+        Storage.set(Timestamp.SCALE_TIMESTAMP_NOW, now.toU8a());
+        return ResponseCodes.SUCCESS;
+    }
+
+    /**
+     *  @description Gets the current time that was set. If this function is called prior 
+     *  to setting the timestamp, it will return the timestamp of the previous block.
+     */
+    static get(): Moment {
+        const now = Storage.get(Timestamp.SCALE_TIMESTAMP_NOW);
+        return now.isSome() ? BytesReader.decodeInto<Moment>((<ByteArray>now.unwrap()).values) : instantiate<Moment>(0);
+    }
+
+    /**
+     * @description Creates timestamp inherent data
+     * @param data inherent data to extract timestamp from
+     */
+    static createInherent(data: IInherentData): Inherent {
+        const timestampData: Moment = BytesReader.decodeInto<Moment>(this.extractInherentData(data).values);
+        let nextTime = timestampData;
+        if(Timestamp.get().value){
+            let nextTimeValue = <u64>(Math.max(<f64>timestampData.value, <f64>(Timestamp.get().value + TimestampTypes.minimumPeriod().value)));
```

Comment:
Aren't we assuming that `Moment` is `UInt64` if we are casting to `u64`? 
Maybe we can implement our own `max` function that compares the 2 variables without casting them to `f64`